Redes Neuronales para Lenguaje Natural, 2025

---
# Laboratorio 2

En este laboratorio construiremos un sistema de Question Answering (QA) utilizando el método de Retrieval-Augmented Generation (RAG), que implica el uso de un paso de recuperación de información y un paso de generación de respuesta con LLM.

**Entrega: 18/11**

**Se debe entregar un archivo zip que contenga:**
* Este notebook de Python (.ipynb) completo.
* Los documentos obtenidos y utilizados como fuentes de información según se explica en la parte 1 (opcionalmente se puede entregar un archivo CSV con los textos de cada documento).
* Archivo CSV con el conjunto de preguntas y respuestas como se explica en la parte 5.

**No olvidar mantener todas las salidas de cada región de código en el notebook!**

---



## Parte 0


### Instalación bibliotecas

In [1]:
if 'google.colab' in str(get_ipython()):
  print('Running on CoLab')
  COLAB = True
else:
  print('Not running on CoLab')
  COLAB = False

#@title Estilo de salida de colab
from IPython.display import HTML, display, clear_output
if COLAB:
    pre_run_cell_fn = lambda: display(HTML('''<style> pre {white-space: pre-wrap;}</style>'''))
    get_ipython().events.register('pre_run_cell', pre_run_cell_fn)

import sys
!{sys.executable} -m pip install transformers -U
!{sys.executable} -m pip install bitsandbytes
!{sys.executable} -m pip install accelerate
!{sys.executable} -m pip install sentence-transformers
!{sys.executable} -m pip install evaluate
!{sys.executable} -m pip install bert_score
!{sys.executable} -m pip install google-genai
!{sys.executable} -m pip install pymupdf4llm
!{sys.executable} -m pip install -qU langchain-text-splitters
# !{sys.executable} -m pip install --upgrade huggingface_hub

clear_output()

### Creación dataset preguntas y respuestas

In [2]:
import os
import csv
TEST_DATASET = "testset.csv"

if not os.path.exists(TEST_DATASET):
    with open(TEST_DATASET, "w", newline="") as csv_file:
        writer = csv.writer(csv_file)
        writer.writerow(["question", "answer", "source"])

        # No relacionadas con el dominio
        respuesta_para_pregunta_fuera_de_tema = "Lo siento, no cuento con información para responder esa pregunta."
        writer.writerow(["¿Dónde nace el río Uruguay?", respuesta_para_pregunta_fuera_de_tema, "N/A"])
        writer.writerow(["¿En qué año se firmó el tratado de Tordesillas?", respuesta_para_pregunta_fuera_de_tema, "N/A"])
        writer.writerow(["¿Cuál es el presidente actual de Chile?", respuesta_para_pregunta_fuera_de_tema, "N/A"])
        writer.writerow(["¿Cuál es la capital de Francia?", respuesta_para_pregunta_fuera_de_tema, "N/A"])
        writer.writerow(["¿Quién ganó la Copa del Mundo de la FIFA 2022?", respuesta_para_pregunta_fuera_de_tema, "N/A"])
        writer.writerow(["¿Cuál es la fórmula química del agua?", respuesta_para_pregunta_fuera_de_tema, "N/A"])

        # Necesitan información de un chunk
        writer.writerow(["¿Es la Biblioteca Nacional más antigua que la república de Uruguay?", "Sí, la Biblioteca Nacional es más antigua que la propia república.", "Presentacion.pdf"])
        writer.writerow(["¿En qué año fue nombrado Francisco Acuña de Figueroa director de la Biblioteca Nacional?", "Fue nombrado a mediados de 1840.", "Acuña de Figueroa.pdf"])
        writer.writerow(["¿Qué director de la Biblioteca Nacional es también considerado el primer poeta de la patria?", "Francisco Acuña de Figueroa.", "Acuña de Figueroa.pdf"])
        writer.writerow(["¿Qué botánico francés visitó la biblioteca cerrada hacia 1820 y estimó su colección?", "Auguste de Saint-Hilaire, quien estimó que el número de libros era de aproximadamente dos mil.", "Clausura, expolios, intentos de reapertura.pdf"])
        writer.writerow(["¿Quién fue el arquitecto que ganó el concurso de 1937 para el actual edificio de la Biblioteca Nacional?", "El arquitecto Luis Crespi.", "El edificio de la Biblioteca Nacional.pdf"])
        writer.writerow(["¿Qué director propuso en 1873 la ""Creación de una Nueva Biblioteca Nacional"" mediante una suscripción popular?", "Juan Antonio Tavolara.", "La Nueva Biblioteca Nacional.pdf"])
        writer.writerow(["¿Qué relación tuvieron Dámaso Antonio Larrañaga y Acuña de Figueroa con la Biblioteca Nacional?", "Dámaso Antonio Larrañaga fue el protagonista del acto fundacional de la biblioteca. Francisco Acuña de Figueroa fue director de la biblioteca durante siete años, en los complejos tiempos del Sitio Grande de Montevideo.", "Presentacion.pdf;Acuña de Figueroa.pdf"])
        writer.writerow(["¿Qué dos directores de la Biblioteca Nacional tuvieron gestiones criticadas o problemáticas?", "La gestión de Francisco Acuña de Figueroa fue polémica por su ""camaleonismo político"" y su fidelidad a Rivera, que llevó a su destitución. La gestión de Juan Antonio Tavolara fue criticada por rematar obras consideradas ""inservibles"" a vil precio y Fernández Saldaña la consideró ""poco brillante"".", "Acuña de Figueroa.pdf;La Nueva Biblioteca Nacional.pdf"])


        # Necesitan información de más de un chunk
        # <<...la construcción de la sede actual...>> <<...se trasladó...>> <<..el alquiler ... que la biblioteca ocupaba desde 1894,31 en la calle Florida n.º 93.. >> <<...no es solo un edificio ... en 18 de julio..>>
        writer.writerow(["¿La Biblioteca Nacional siempre estuvo en su lugar actual?", "No, la sede de la Biblioteca Nacional ha cambiado varias veces a lo largo de la historia.", "Presentacion.pdf, La Nueva Biblioteca Nacional.pdf, El edificio de la Biblioteca Nacional.pdf"])

        # <<...actualmente cuenta con 850.000 volúmenes...>> <<..entre 1916 y 1933 se pasó de 59.552 a 132.442 volúmenes...>>
        writer.writerow(["¿En que año la Biblioteca Nacional contó con más volúmenes?", "En 2021, al alcanzar los 850.000 volúmenes", "La Nueva Biblioteca Nacional.pdf, El edificio de la Biblioteca Nacional.pdf"])

        # <<..fue director por 7 años...>> <<..cuando a mediados de 1840... decidió conferirle el cargo de director...>>
        writer.writerow(["¿En que año terminó el mandato de Francisco Acuña de Figueroa como director de la Biblioteca Nacional?", "En 1847", "Acuña de Figueroa.pdf"])


## Parte 1: Procesamiento de los documentos

En esta parte, cada grupo deberá construir y procesar su conjunto de documentos. Esto consiste de los siguientes pasos:

* Elegir un tema dentro de un dominio específico sobre el que trabajar.
* Obtener al menos 5 documentos en español que contengan información sobre el tema elegido.
* Procesar cada documento para extraer el texto del formato original a un string en Python (por ejemplo, extraer el texto de un PDF).

El resultado de esta parte debe ser una lista cargada en memoria que contenga el texto (string) de cada uno de los documentos elegidos.

**Sugerencias:**
* Se recomienda utilizar artículos de wikipedia para simplificar la etapa de extracción del texto (ver la librería [wikipedia-api](https://github.com/martin-majlis/Wikipedia-API/)).
* Opcionalmente puede utilizar documentos PDF, páginas web u otros formatos. En estos casos se sugiere:
  * Utilizar la librería PyPDF2 para procesar documentos PDF.
  * Utilizar la librería LangChain para procesar páginas web, en particular la clase Html2TextTransformer, que convierte HTML a Markdown ([ejemplo de uso](https://python.langchain.com/v0.2/docs/integrations/document_transformers/html2text/)).
* Puede ser conveniente guardar el resultado del procesamiento de los documentos en un archivo CSV (donde cada fila corresponde al texto de un documento) para no tener que repetir este proceso cada vez que se ejecuta el notebook, y en su lugar cargar el archivo CSV.

Usé PyMuPDF en vez de PyPDF2 para poder pasar el pdf a markdown.

Duda -> No sé si usar la página web o el pdf, la página web no tiene todo el documento cargado a priori, hay que scrollear en el iFrame para que aparezca en el html, no sé si suma tanta data tampoco

TO DO -> Revisar como manejar las citas / notas al pie. Enriquecimiento de texto etc

In [3]:
import pymupdf4llm
import csv

CSV_NAME = "corpus.csv"

def extractTextFromPdf(file_name):
    markdown = pymupdf4llm.to_markdown(file_name)
    return markdown

def writeIntoCsv(file_name, text):
    with open(CSV_NAME, "a", newline="") as csv_file:
        writer = csv.writer(csv_file)
        writer.writerow([file_name, text])

def addFileToCsv(file_name):
    text = extractTextFromPdf(file_name)
    writeIntoCsv(file_name, text)

def createCsv():
    with open(CSV_NAME, "w", newline="") as csv_file:
        writer = csv.writer(csv_file)
        writer.writerow(["file_name", "text"])


Consider using the pymupdf_layout package for a greatly improved page layout analysis.


In [4]:
files = ["Presentacion.pdf", "Acuña de Figueroa.pdf", "Clausura, expolios, intentos de reapertura.pdf", "El edificio de la Biblioteca Nacional.pdf", "La Nueva Biblioteca Nacional.pdf"]

createCsv()
for file_name in files:
  addFileToCsv(file_name)

Los textos resultantes deben estar almacenados en la variable `documents`:

In [5]:
import pandas as pd
documents = pd.read_csv(CSV_NAME, header = 0, index_col='file_name')

In [6]:
print(documents)

                                                                                             text
file_name                                                                                        
Presentacion.pdf                                Revista de la Biblioteca Nacional. 17, 5, 2021...
Acuña de Figueroa.pdf                           Revista de la Biblioteca Nacional. 17, 71-90, ...
Clausura, expolios, intentos de reapertura.pdf  Revista de la Biblioteca Nacional. 17, 53-69, ...
El edificio de la Biblioteca Nacional.pdf       Revista de la Biblioteca Nacional. 17, 225-265...
La Nueva Biblioteca Nacional.pdf                Revista de la Biblioteca Nacional. 17, 109-117...


## Parte 2: Chunking

Una vez que se obtiene el texto de cada documento, se debe realizar la etapa de _chunking_. Esta etapa consiste en dividir cada texto en segmentos más chicos a los que llamamos _chunks_.

Realizar la etapa de _chunking_ de forma automática utilizando un método simple que permita obtener _chunks_ de un largo aproximado de 500 caracteres.

Puede probar con dividir a nivel de caracteres, palabras o incluso párrafos, teniendo en cuenta que el largo de cada _chunk_ no debería exceder demasiado los 500 caracteres.

**Sugerencias:**
* Puede utilizar los splitters disponibles en LangChain ([documentación](https://python.langchain.com/v0.1/docs/modules/data_connection/document_transformers/)) como RecursiveCharacterTextSplitter, aunque no es obligatorio y también es correcto hacer una implementación propia.
* Tener en cuenta que esta etapa es crucial en el resultado final. Cuanto más contextualizados queden los *chunks*, mejor será el rendimiento de la etapa de recuperación de información. Es conveniente minimizar la división de palabras (o párrafos) por la mitad.

#### Definición de splitters

In [7]:
from langchain_text_splitters import RecursiveCharacterTextSplitter
from langchain_text_splitters import MarkdownHeaderTextSplitter

from enum import StrEnum

class SplitterType(StrEnum):
    RECURSIVE = "Recursive"
    MARKDOWN = "Markdown"

## Recursive Splitter
text_splitter = RecursiveCharacterTextSplitter(
    chunk_size=450,
    chunk_overlap=200, # Con overlaps más chicos el splitter ignora el parámetro
    length_function=len,
    is_separator_regex=False
)

def splitRecursivo(text):
  chunks = text_splitter.split_text(text)
  return chunks

## Markdown Splitter
headers_to_split_on = [
    ("#", "Header 1"),
    ("##", "Header 2"),
    ("###", "Header 3"),
    ("####", "Header 4"),
]

markdown_splitter = MarkdownHeaderTextSplitter(
    headers_to_split_on,
    return_each_line=True
)

def splitMarkdown(text):
  markdown_chunks = markdown_splitter.split_text(text)
  chunks = text_splitter.split_documents(markdown_chunks)
  return chunks

In [8]:
# incorpora la jerarquía de headers al texto base en el caso del Markdown Splitter
def agregar_titulos(chunks):
    for chunk in chunks:
      header_string = "\n".join(chunk.metadata.values())
      final_string = f"{header_string}\n{chunk.page_content}"
      chunk.page_content = final_string
    return chunks

In [9]:
def chunk_text(text, splitter=SplitterType.RECURSIVE):

  if splitter == SplitterType.RECURSIVE:
    chunks = splitRecursivo(text)
  elif splitter == SplitterType.MARKDOWN:
    chunks = splitMarkdown(text)
    chunks = agregar_titulos(chunks)

  return chunks # Lista de strings con los chunks del texto

In [10]:
def chunk_documents(chunk_list, splitter=SplitterType.RECURSIVE):
  for document_text in documents['text']:
      chunk_list += chunk_text(document_text,splitter=splitter)
  if splitter == SplitterType.RECURSIVE:
        print("Documents chunked using recursive splitter")
  elif splitter == SplitterType.MARKDOWN:
        print("Documents chunked using markdown splitter")
  return chunk_list

In [11]:
recursive_chunks = []
recursive_chunks = chunk_documents(recursive_chunks, splitter=SplitterType.RECURSIVE)
markdown_chunks = []
markdown_chunks = chunk_documents(markdown_chunks, splitter=SplitterType.MARKDOWN)

Documents chunked using recursive splitter
Documents chunked using markdown splitter


### Codigo para ver chunks resultantes

In [12]:
#Recursivo
print("\n--- Algunos chunks recursivos: ---")
for i, chunk in enumerate(recursive_chunks[:4]):
    print(f"--- CHUNK {i+1} (Length: {len(chunk)}) ---")
    print(chunk)
    print("\n")


--- Algunos chunks recursivos: ---
--- CHUNK 1 (Length: 253) ---
Revista de la Biblioteca Nacional. 17, 5, 2021. ISSN 0797-9061

# **Presentación**


La Biblioteca Nacional no es solo un edificio solemne en la ave­
nida 18 de Julio. Es también una historia que merece ser recordada
y contada a las nuevas generaciones.


--- CHUNK 2 (Length: 309) ---
Esa historia encierra una épica silenciosa: hace falta mucha tena­
cidad, mucho valor y mucho compromiso con la cultura para poder
tener una biblioteca pública fundada hace 205 años, en un país que
todavía no ha celebrado su segundo centenario. La Biblioteca Nacio­
nal es más antigua que la propia república.


--- CHUNK 3 (Length: 399) ---
Esta entrega repasa las diferentes etapas de ese trayecto, desde
el acto fundacional protagonizado por Dámaso Antonio Larrañaga
hasta este hoy marcado por profundas tendencias de cambio cultural
y tecnológico. Ente ambos extremos desfilan grandes nombres (Acu­
ña de Figueroa y José Enrique Rodó, entre otr

In [13]:
#Markdown
print("\n--- Algunos chunks markdown: ---")
for i, chunk in enumerate(markdown_chunks[:4]):
    print(f"\n--- CHUNK {i+1} (Length: {len(chunk.page_content)}) ---")
    print(chunk)
    print("\n")


--- Algunos chunks markdown: ---

--- CHUNK 1 (Length: 63) ---
page_content='
Revista de la Biblioteca Nacional. 17, 5, 2021. ISSN 0797-9061'



--- CHUNK 2 (Length: 184) ---
page_content='**Presentación**
La Biblioteca Nacional no es solo un edificio solemne en la ave
nida 18 de Julio. Es también una historia que merece ser recordada
y contada a las nuevas generaciones.' metadata={'Header 1': '**Presentación**'}



--- CHUNK 3 (Length: 324) ---
page_content='**Presentación**
Esa historia encierra una épica silenciosa: hace falta mucha tena
cidad, mucho valor y mucho compromiso con la cultura para poder
tener una biblioteca pública fundada hace 205 años, en un país que
todavía no ha celebrado su segundo centenario. La Biblioteca Nacio
nal es más antigua que la propia república.' metadata={'Header 1': '**Presentación**'}



--- CHUNK 4 (Length: 415) ---
page_content='**Presentación**
Esta entrega repasa las diferentes etapas de ese trayecto, desde
el acto fundacional protagonizado por 

## Parte 3: Recuperación de información

En esta parte vamos a implementar el método de recuperación de información que nos permitirá obtener los _chunks_ más relevantes para la pregunta.

En primer lugar, cargamos el modelo Bi-Encoder que utilizaremos para generar los embeddings utilizando la librería sentence_transformers.

Se utiliza el modelo multilingüe [intfloat/multilingual-e5-large](https://huggingface.co/intfloat/multilingual-e5-large), fine-tuning del modelo `xlm-roberta-large` para la tarea de generación de sentence embeddings.

Se pueden explorar otros modelos Bi-Encoder, e incluso modelos Cross-Encoder o del tipo ColBERT. En HuggingFace se puede consultar el siguiente [leaderboard](https://huggingface.co/spaces/mteb/leaderboard) que compara varios modelos de este tipo en diferentes tareas.

In [14]:
from sentence_transformers import SentenceTransformer

model_emb = SentenceTransformer("intfloat/multilingual-e5-large")

modules.json:   0%|          | 0.00/387 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/57.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/690 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/2.24G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/418 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.1M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/280 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/201 [00:00<?, ?B/s]

A continuación se debe generar las representaciones vectoriales para todos los _chunks_ ([ejemplo de uso](https://huggingface.co/intfloat/multilingual-e5-large#support-for-sentence-transformers)).

**Observación:** El modelo que estamos usando espera que los _chunks_ comiencen con el prefijo `passage: ` por lo que será necesario agregarlo al inicio de todos los _chunks_.

In [15]:
def prepare_chunks_for_embedding(chunks):
    prepared_chunks = []
    i = 0
    for chunk in chunks:
        # Si es un objeto Document (del markdown splitter), extraer el contenido
        if hasattr(chunk, 'page_content'):
            text = chunk.page_content
        else:
            # Si es un string simple
            text = chunk

        prepared_chunks.append(f"passage: {text}")

    return prepared_chunks

chunks_with_prefix_recursive = prepare_chunks_for_embedding(recursive_chunks)
chunks_with_prefix_markdown = prepare_chunks_for_embedding(markdown_chunks)

print(f"Generando embeddings para {len(chunks_with_prefix_recursive)} chunks recursivos...")
chunk_embeddings_recursive = model_emb.encode(chunks_with_prefix_recursive, normalize_embeddings=True)

print(f"Generando embeddings para {len(chunks_with_prefix_markdown)} chunks markdown...")
chunk_embeddings_markdown = model_emb.encode(chunks_with_prefix_markdown, normalize_embeddings=True)

print(f"Embeddings recursivos generados. Dimensión: {chunk_embeddings_recursive.shape}")
print(f"Embeddings markdown generados. Dimensión: {chunk_embeddings_markdown.shape}")

Generando embeddings para 503 chunks recursivos...
Generando embeddings para 742 chunks markdown...
Embeddings recursivos generados. Dimensión: (503, 1024)
Embeddings markdown generados. Dimensión: (742, 1024)


Por último, se debe implementar el algoritmo de búsqueda de los embeddings más cercanos para un embedding dado.

**Sugerencias:**
* Utilizar la clase NearestNeighbors de sklearn ([documentación](https://scikit-learn.org/dev/modules/generated/sklearn.neighbors.NearestNeighbors.html#sklearn.neighbors.NearestNeighbors)).

In [16]:
from sklearn.neighbors import NearestNeighbors

nn_model_recursive = NearestNeighbors(
    n_neighbors=5,  # Por ahora buscamos 5 vecinos, ajustamos despues
    metric='cosine',
    algorithm='brute'  # 'brute' supuestamente es más preciso para datasets chicos/medianos
)

nn_model_recursive.fit(chunk_embeddings_recursive)

print("Modelo de vecinos más cercanos recursivo entrenado")

nn_model_markdown = NearestNeighbors(
    n_neighbors=5,  # Por ahora buscamos 5 vecinos, ajustamos despues
    metric='cosine',
    algorithm='brute'  # 'brute' supuestamente es más preciso para datasets chicos/medianos
)

nn_model_markdown.fit(chunk_embeddings_markdown)

print("Modelo de vecinos más cercanos markdown entrenado")

Modelo de vecinos más cercanos recursivo entrenado
Modelo de vecinos más cercanos markdown entrenado


In [17]:
def retrieve_chunks(all_chunks, nn_model, query, top_k=3):
    """
    Returns:
        Una tupla (chunks_recuperados, distancias, indices)
        - chunks_recuperados: Lista de textos de los chunks más relevantes
        - distancias: Distancias coseno a cada chunk
        - indices: Índices de los chunks en la lista original
    """
    query_with_prefix = f"query: {query}"

    query_embedding = model_emb.encode([query_with_prefix], normalize_embeddings=True)

    distances, indices = nn_model.kneighbors(query_embedding, n_neighbors=top_k)

    # Extraer los chunks correspondientes
    retrieved_chunks = []
    for idx in indices[0]:
        chunk = all_chunks[idx]
        # Manejar tanto strings como objetos Document
        if hasattr(chunk, 'page_content'):
            retrieved_chunks.append(chunk.page_content)
        else:
            retrieved_chunks.append(chunk)

    return retrieved_chunks, distances[0], indices[0]

## Parte 4: Generación de respuestas

### Configuración de LLM

En esta parte, implementaremos un wrapper flexible que permite experimentar con diferentes modelos de lenguaje:

1. **Llama 3.1** (modelo abierto): Utilizaremos el modelo Meta-Llama-3.1-8B-Instruct a través de HuggingFace.
2. **Gemini 2.0 Flash** (modelo cerrado): Utilizaremos la API de Google Gemini.

Para **Llama 3.1**, es necesario:
- Crearse una cuenta de HuggingFace (https://huggingface.co/)
- Aceptar los términos para usar el modelo en HuggingFace: https://huggingface.co/meta-llama/Meta-Llama-3.1-8B-Instruct
- Crear un token de HuggingFace con permiso de lectura: https://huggingface.co/settings/tokens

Para **Gemini**, es necesario:
- Obtener una API key de Google AI Studio: https://aistudio.google.com/app/apikey

In [18]:
# Ejecutar para conectarse a HuggingFace (solo necesario para Llama 3.1)
from huggingface_hub import notebook_login, login

if COLAB:
    notebook_login()
else:
    token = ''
    login(token=token)


### Definición del Wrapper de LLM

A continuación se define una clase abstracta que permite intercambiar fácilmente entre diferentes modelos de lenguaje.

In [19]:
from abc import ABC, abstractmethod
from typing import Optional
import os
from getpass import getpass

class LLMWrapper(ABC):
    """Clase base abstracta para wrappers de modelos de lenguaje."""

    @abstractmethod
    def generate(self, prompt: str, temperature: float = 0.0, max_tokens: int = 500) -> str:
        pass

    @abstractmethod
    def get_model_name(self) -> str:
        pass


class LlamaWrapper(LLMWrapper):
    """Wrapper para el modelo Llama 3.1 de Meta vía HuggingFace."""

    def __init__(self):
        from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig
        import torch

        print("Inicializando Llama 3.1...")

        self.tokenizer = AutoTokenizer.from_pretrained(
            "meta-llama/Meta-Llama-3.1-8B-Instruct")


        # Inicializar el modelo
        if COLAB:
                    # Configuración de cuantización a 4 bits (para mejorar eficiencia)
            bnb_config = BitsAndBytesConfig(
                load_in_4bit=True,
                bnb_4bit_quant_type="nf4",
                bnb_4bit_compute_dtype=torch.bfloat16)


            self.model = AutoModelForCausalLM.from_pretrained(
                "meta-llama/Meta-Llama-3.1-8B-Instruct",
                quantization_config=bnb_config,
                device_map="auto")
        else:
            self.model = AutoModelForCausalLM.from_pretrained(
                "meta-llama/Meta-Llama-3.1-8B-Instruct",
                device_map="auto")

        print("Llama 3.1 inicializado correctamente")

    def generate(self, prompt: str, temperature: float = 0.0, max_tokens: int = 500) -> str:
        from transformers import GenerationConfig, pipeline

        # Configuración de temperatura
        generation_config = GenerationConfig(
            temperature=temperature if temperature > 0 else None,
            do_sample=temperature > 0)

        # Inicializar pipeline para generación de texto
        pipe = pipeline(
            "text-generation",
            model=self.model,
            config=generation_config,
            tokenizer=self.tokenizer,
            pad_token_id=self.tokenizer.eos_token_id)

        # Generar texto
        output = pipe(
            prompt,
            return_full_text=False,
            max_new_tokens=max_tokens)

        return output[0]['generated_text']

    def get_model_name(self) -> str:
        return "Llama-3.1-8B-Instruct"


class GeminiWrapper(LLMWrapper):
    """Wrapper para el modelo Gemini 2.0 Flash de Google."""

    def __init__(self, api_key: Optional[str] = None):
        from google import genai

        print("Inicializando Gemini 1.5 Flash...")

        # Obtener API key
        if api_key is None:
            api_key = os.environ.get("GEMINI_API_KEY")

        if api_key is None:
            print("No se encontró la API key de Gemini en las variables de entorno.")
            api_key = getpass("Por favor, ingrese su API key de Gemini: ")
            # Guardar en variables de entorno para esta sesión
            os.environ["GEMINI_API_KEY"] = api_key

        self.client = genai.Client(api_key=api_key)

        print("Gemini 2.0 Flash inicializado correctamente")

    def generate(self, prompt: str, temperature: float = 0.0, max_tokens: int = 500) -> str:
        from google.genai import types

        config = types.GenerateContentConfig(
            temperature=temperature if temperature > 0 else 0.0,
            max_output_tokens=max_tokens)

        # Generar respuesta
        response = self.client.models.generate_content(
            model="gemini-2.0-flash",
            contents=prompt,
            config=config)

        return response.text

    def get_model_name(self) -> str:
        return "Gemini-2.0-Flash"

### Instanciar modelos

Seleccionar qué modelo(s) se va a inicializar. Se puede inicializar ambos para facilitar la experimentación posterior.

In [20]:
# Descomentar el(los) modelo(s) que se quiera utilizar

llama_model = LlamaWrapper()

gemini_model = GeminiWrapper()

# Seleccionar el modelo activo para los experimentos manuales
active_model = llama_model

print(f"\nModelo activo: {active_model.get_model_name()}")

Inicializando Llama 3.1...


tokenizer_config.json:   0%|          | 0.00/55.4k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/9.09M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/296 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/855 [00:00<?, ?B/s]

model.safetensors.index.json:   0%|          | 0.00/23.9k [00:00<?, ?B/s]

Fetching 4 files:   0%|          | 0/4 [00:00<?, ?it/s]

model-00003-of-00004.safetensors:   0%|          | 0.00/4.92G [00:00<?, ?B/s]

model-00002-of-00004.safetensors:   0%|          | 0.00/5.00G [00:00<?, ?B/s]

model-00004-of-00004.safetensors:   0%|          | 0.00/1.17G [00:00<?, ?B/s]

model-00001-of-00004.safetensors:   0%|          | 0.00/4.98G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/4 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/184 [00:00<?, ?B/s]

Llama 3.1 inicializado correctamente
Inicializando Gemini 1.5 Flash...
No se encontró la API key de Gemini en las variables de entorno.
Por favor, ingrese su API key de Gemini: ··········
Gemini 2.0 Flash inicializado correctamente

Modelo activo: Llama-3.1-8B-Instruct


### Función auxiliar para generación de respuestas

Esta función utiliza el modelo activo seleccionado anteriormente.

In [21]:
def get_response(
    prompt: str,
    model: LLMWrapper = None,
    temp: float = 0.0,
    max_tok: int = 500
) -> str:
    if model is None:
        model = active_model

    return model.generate(prompt, temperature=temp, max_tokens=max_tok)

### Crear prompt y generar respuesta

Escribir la función `create_prompt(question, use_chat_template=True)` que dada una pregunta, genere la prompt que se utilizará para generar la respuesta. Tener en cuenta que se debe realizar la búsqueda semántica de los _chunks_ más cercanos a la pregunta utilizando lo implementado en la parte 3.

**Observación:** Al igual que para los _chunks_, el modelo Bi-Encoder espera que la pregunta comience con un prefijo especial: `query: ` por lo que será necesario agregarlo al inicio de la pregunta para generar el embedding.

**Sugerencias:**
* Puede probar con distintas cantidades de _chunks_ recuperados, pero se sugiere comenzar con 3. Tener en cuenta que más _chunks_ recuperados y agregados en la prompt implica mayor uso de memoria en inferencia.
* El parámetro `use_chat_template` permite controlar si se aplica el template de chat de Llama (necesario para Llama 3.1, opcional para Gemini). Para Llama usar `True`, para Gemini se puede probar con `True` o `False` según el formato que se prefiera.

In [22]:
def create_prompt(question, use_chat_template=True, model_for_template=None, top_k=3, chunk_list=recursive_chunks, nn_model=nn_model_recursive):
    """
    Crea el prompt para el modelo de lenguaje incluyendo contexto recuperado.

    Args:
        question: La pregunta del usuario
        use_chat_template: Si True, aplica el template de chat de Llama
        model_for_template: Modelo del cual usar el tokenizer (solo para Llama)
        top_k: Número de chunks a recuperar (por defecto 3)

    Returns:
        El prompt formateado
    """
    # 1. Recuperar chunks relevantes usando búsqueda semántica
    # La función retrieve_chunks ya maneja el prefijo "query:" internamente
    retrieved_chunks, _, _ = retrieve_chunks(chunk_list, nn_model, question, top_k=top_k)

    # 2. Construir el contexto con los chunks recuperados
    context = "\n\n".join(retrieved_chunks)

    # 3. Crear el mensaje del sistema y del usuario
    system_message = """Eres un asistente experto en responder preguntas basándote únicamente en el contexto proporcionado.

Instrucciones:
- Si la información para responder la pregunta está en el contexto, proporciona una respuesta clara y concisa.
- Si la información NO está en el contexto, responde: "Lo siento, no cuento con información para responder esa pregunta."
- No inventes información que no esté en el contexto.
- Responde en español."""

    user_message = f"""Contexto:
{context}

Pregunta: {question}"""

    # 4. Aplicar chat template si corresponde
    if use_chat_template:
        # Usar tokenizer de Llama para aplicar template
        if model_for_template is None and isinstance(active_model, LlamaWrapper):
            model_for_template = active_model

        if isinstance(model_for_template, LlamaWrapper):
            messages = [
                {"role": "system", "content": system_message},
                {"role": "user", "content": user_message}
            ]
            prompt = model_for_template.tokenizer.apply_chat_template(
                messages, tokenize=False, add_generation_prompt=True
            )
        else:
            # Para Gemini u otros modelos sin template específico
            prompt = f"{system_message}\n\n{user_message}"
    else:
        # Sin template, formato simple
        prompt = f"{system_message}\n\n{user_message}"

    return prompt

Probar la prompt anterior con un ejemplo.

In [ ]:
question = "¿En que año terminó el mandato de Francisco Acuña de Figueroa como director de la Biblioteca Nacional?"  # Completar con una pregunta adecuada al contexto

# Crear prompt según el modelo activo
use_template = isinstance(active_model, LlamaWrapper)  # True para Llama, False para Gemini
prompt = create_prompt(recursive_chunks, nn_model_recursive, question, use_chat_template=use_template)

print(f"MODELO: {active_model.get_model_name()}")
print("\nPROMPT:")
print(prompt)
print("\nRESPUESTA:")
print(get_response(prompt))

## Parte 5: Evaluación
A continuación vamos a evaluar la solución construida. Para ello, se deben seguir los siguientes pasos:

* Construir un conjunto de evaluación de forma manual que contenga al menos 12 preguntas y respuestas con las siguientes características:
  * Al menos 3 preguntas deben necesitar información presente en más de un _chunk_ para ser respondidas correctamente.
  * Al menos 3 preguntas no deben estar relacionadas con el dominio, y su respuesta de referencia debe ser algo similar a: "Lo siento, no cuento con información para responder esa pregunta."
* El conjunto debe estar en un archivo CSV llamado testset.csv, con las columnas "question" y "answer".

Se deberá realizar al menos tres experimentos diferentes y evaluar sobre el mismo conjunto de test con la métrica BERTScore. Los experimentos deben variar en al menos uno de los siguientes elementos:
* Método de chunking
* Modelo (o método) de retrieval
* Modelo de generación (LLM)
* Método de prompting (se puede probar con few-shot, chain of thought, etc)
* Otros aspectos que considere relevantes a probar

A continuación se definen funciones auxiliares para la evaluación.


In [28]:
import evaluate
import numpy as np
import time
from tqdm.notebook import tqdm

def generate_predictions(
    questions,
    model=None,
    use_chat_template=True,
    use_delay=False,
    chunk_list=recursive_chunks,
    nn_model=nn_model_recursive
):
    if model is None:
        model = active_model

    predictions = []
    for question in tqdm(questions, desc=f"Generando con {model.get_model_name()}"):
        prompt = create_prompt(question, use_chat_template=use_chat_template, chunk_list=chunk_list, nn_model=nn_model)
        prediction = get_response(prompt, model=model)
        if use_delay:
            time.sleep(3)
        predictions.append(prediction)

    return predictions

def evaluate_predictions(predictions, references, experiment_name=""):
    """Evalúa predicciones usando BERTScore."""
    bertscore = evaluate.load("bertscore")
    results = bertscore.compute(predictions=predictions, references=references, lang='es')

    metrics = {
        'precision': np.array(results['precision']).mean(),
        'recall': np.array(results['recall']).mean(),
        'f1': np.array(results['f1']).mean()
    }

    if experiment_name:
        print(f"\n=== Resultados: {experiment_name} ===")
    print(f"BERTScore P:  {metrics['precision']:.3f}")
    print(f"BERTScore R:  {metrics['recall']:.3f}")
    print(f"BERTScore F1: {metrics['f1']:.3f}")

    return metrics

In [24]:
import pandas as pd

# Leer el conjunto de evaluación
df = pd.read_csv("testset.csv")

# Obtener preguntas y respuestas
questions = df["question"].tolist()
references = df["answer"].tolist()

Evalúe los experimentos realizados.

In [25]:
import json
from google.colab import files

In [29]:
# Evaluar experimentos
# Almacenar resultados para comparación posterior
results_dict = {}

# ============================================================================
# EXPERIMENTO 1: Llama 3.1 con chunking recursivo
# ============================================================================
exp1_model = llama_model
exp1_name = "Exp1: Llama 3.1 con chunking recursivo"

predictions_exp1 = generate_predictions(questions, model=exp1_model, use_chat_template=True, chunk_list=recursive_chunks, nn_model=nn_model_recursive)
results_dict[exp1_name] = evaluate_predictions(predictions_exp1, references, exp1_name)

##saving to file
with open('exp1.json', 'w', encoding='utf-8') as f:
    json.dump(predictions_exp1, f, indent=2)
    json.dump(results_dict[exp1_name], f, indent=2)

files.download('exp1.json')
##end saving to file

# ============================================================================
# EXPERIMENTO 1: Llama 3.1 con chunking markdown
# ============================================================================
exp2_model = llama_model
exp2_name = "Exp2: Llama 3.1 con chunking markdown"

predictions_exp2 = generate_predictions(questions, model=exp2_model, use_chat_template=True, chunk_list=markdown_chunks, nn_model=nn_model_markdown)
results_dict[exp2_name] = evaluate_predictions(predictions_exp2, references, exp2_name)

##saving to file
with open('exp2.json', 'w', encoding='utf-8') as f:
    json.dump(predictions_exp2, f, indent=2)
    json.dump(results_dict[exp2_name], f, indent=2)

files.download('exp2.json')
##end saving to file

# ============================================================================
# EXPERIMENTO 3: Gemini 2.0 Flash con chunking markdown
# ============================================================================
exp3_model = gemini_model
exp3_name = "Exp3: Gemini 2.0 Flash con chunking markdown"

predictions_exp3 = generate_predictions(questions, model=exp3_model, use_chat_template=False, use_delay=True, chunk_list=markdown_chunks, nn_model=nn_model_markdown)
results_dict[exp3_name] = evaluate_predictions(predictions_exp3, references, exp3_name)

##saving to file
with open('exp3.json', 'w', encoding='utf-8') as f:
    json.dump(predictions_exp3, f, indent=2)
    json.dump(results_dict[exp3_name], f, indent=2)

files.download('exp3.json')
##end saving to file


# ============================================================================
# RESUMEN DE RESULTADOS
# ============================================================================
print("\n" + "="*80)
print("RESUMEN DE TODOS LOS EXPERIMENTOS")
print("="*80)

import pandas as pd
summary_df = pd.DataFrame(results_dict).T
summary_df.columns = ['Precision', 'Recall', 'F1']
print(summary_df.to_string())

Generando con Llama-3.1-8B-Instruct:   0%|          | 0/17 [00:00<?, ?it/s]

Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0


tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/625 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/996k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.96M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/714M [00:00<?, ?B/s]


=== Resultados: Exp1: Llama 3.1 con chunking recursivo ===
BERTScore P:  0.810
BERTScore R:  0.827
BERTScore F1: 0.816


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Generando con Llama-3.1-8B-Instruct:   0%|          | 0/17 [00:00<?, ?it/s]

Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0
Device set to use cuda:0



=== Resultados: Exp2: Llama 3.1 con chunking markdown ===
BERTScore P:  0.850
BERTScore R:  0.850
BERTScore F1: 0.849


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Generando con Gemini-2.0-Flash:   0%|          | 0/17 [00:00<?, ?it/s]


=== Resultados: Exp3: Gemini 2.0 Flash con chunking markdown ===
BERTScore P:  0.835
BERTScore R:  0.865
BERTScore F1: 0.849


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>


RESUMEN DE TODOS LOS EXPERIMENTOS
                                              Precision    Recall        F1
Exp1: Llama 3.1 con chunking recursivo         0.809582  0.826782  0.816426
Exp2: Llama 3.1 con chunking markdown          0.849967  0.850140  0.849045
Exp3: Gemini 2.0 Flash con chunking markdown   0.835447  0.865094  0.849232


### Reporte de resultados

Reportar los resultados obtenidos en los experimentos realizados completando la siguiente tabla:

| Exp | Descripción | P BERTScore | R BERTScore | F BERTScore |
|-----|-------------|-------------|-------------|-------------|
| 1 | Llama 3.1 RecursiveSplitter| 0.811| 0.818| 0.812|
| 2 | Llama 3.1 MarkdownSplitter| 0.816|0.818 |0.816 |
| 3 | | | | |

Responda las siguientes preguntas:

1. Explique brevemente las diferencias en los experimentos realizados, ¿Qué aspectos se varió en el pipeline de RAG?

2. ¿Son consistentes los resultados obtenidos con lo que esperaba?

3. ¿Le parece que la métrica BERTScore está capturando correctamente las diferencias de los distintos experimentos realizados?

**Pregunta 1**

Se realizaron variaciones en el método de chunking y el modelo generativo. No se probaron variaciones en el modelo de embeddings (multilingual-e5-large) ni el algoritmo de recuperación de información (k-NN con k=5). Se buscó variar solo un aspecto a la vez para poder evaluar el efecto de éste sobre la performance dentro de lo posible (no tenemos control sobre el no-determinismo propio de los LLMs).
1. En el primer experimento se utilizó el splitter Recursivo de Langchain para realizar el chunking y Llama 3.1 como modelo generativo.
2. En el segundo experimento se utilizó el splitter Markdown de Langchain y luego nuevamente Llama 3.1 como modelo generativo.
3. Finalmente, en el último experimento se mantuvo el splitter <"INSERT HERE (pensaba usar el que haya sido mejor de los otros 2)"> y se utilizó Gemini-2.0-Flash como modelo generativo.


#### Evaluación Humana